File: tests/boundary/set.js
/**
 * Set functions — boundary specs (DoD gate 7, at the boundary).
 * ==============================================================
 *
 * `cargo test` covers the fourteen functions over an ordinary Rust set. What
 * it cannot cover is anything about the JavaScript `Set` objects themselves,
 * and for this unit that is where most of the risk lives:
 *
 *   - The four mutating functions must mutate the CALLER'S object. `test/set.js`
 *     checks the contents afterwards, which a port that replaced the object
 *     would also pass if it were somehow able to; more usefully, it never
 *     checks *how* the object was mutated, and clear-and-rebuild is observably
 *     different from add/delete to anything already iterating it.
 *   - The returned values must be real `Set`s, not lookalikes. `Array.from`
 *     works on both.
 *   - Iteration order is asserted by eight of the fourteen upstream blocks, but
 *     always on inputs where the interesting orderings coincide. The two that
 *     do not — `intersection` following its SMALLEST argument, and ties going
 *     to the first — are unreachable from the original file.
 *   - SameValueZero membership (`NaN`, `-0`) is never exercised.
 *
 * Everything about *values* is asserted differentially against the vendored
 * upstream source in `bench/upstream/`, so a passing spec says "the port and
 * upstream agree" rather than "the port matches what I typed".
 */
var assert = require('assert'),
    fs = require('fs'),
    Module = require('module'),
    path = require('path');

process.env.NODE_PATH = path.resolve(__dirname, '..', 'node_modules') +
  (process.env.NODE_PATH ? path.delimiter + process.env.NODE_PATH : '');
Module._initPaths();

function repositoryRoot() {
  var directory = __dirname;

  for (var i = 0; i < 8; i++) {
    if (fs.existsSync(path.join(directory, 'bench', 'upstream', 'set.js')))
      return directory;

    directory = path.dirname(directory);
  }

  throw new Error('cannot locate bench/upstream from ' + __dirname);
}

var upstream = require(path.join(repositoryRoot(), 'bench', 'upstream', 'set.js')),
    port = require('../set.js');

/**
 * Run `scenario` against both implementations and assert they agree.
 * Returns the agreed value so the caller can also say what it should be.
 */
function agree(name, scenario) {
  var mine = scenario(port),
      theirs = scenario(upstream);

  assert.deepStrictEqual(mine, theirs, name + ': port and upstream disagree');

  return mine;
}

describe('set — boundary', function() {

  describe('the returned set', function() {

    it('should be a real Set, not something Array.from happens to accept.', function() {
      var A = new Set([1, 2, 3]),
          B = new Set([2, 3, 4]);

      ['intersection', 'union', 'difference', 'symmetricDifference'].forEach(function(name) {
        var result = port[name](A, B);

        assert.ok(result instanceof Set, name + ' must return a Set');
        assert.strictEqual(result.constructor, Set);
      });
    });

    it('should be a fresh set even when difference short-circuits.', function() {
      // `difference(A, B)` with an empty B returns `new Set(A)` upstream -- a
      // COPY. Returning A itself would pass every assertion in test/set.js and
      // alias two variables the caller believes are independent.
      var A = new Set([1, 2]),
          empty = new Set();

      var mine = port.difference(A, empty),
          theirs = upstream.difference(A, empty);

      assert.notStrictEqual(mine, A);
      assert.notStrictEqual(theirs, A);
      assert.deepStrictEqual(Array.from(mine), Array.from(theirs));
    });
  });

  describe('iteration order', function() {

    it('should follow the SMALLEST argument in intersection.', function() {
      // Unreachable from test/set.js: its two-set case uses equal sizes and
      // its variadic case intersects down to a single member.
      var order = agree('intersection order', function(f) {
        return Array.from(f.intersection(new Set([3, 2, 1]), new Set([1, 2])));
      });

      assert.deepStrictEqual(order, [1, 2]);
    });

    it('should break a size tie in favour of the first argument.', function() {
      var order = agree('intersection tie', function(f) {
        return Array.from(f.intersection(new Set([3, 2, 1]), new Set([1, 2, 3])));
      });

      assert.deepStrictEqual(order, [3, 2, 1]);
    });

    it('should put A\'s half before B\'s in symmetricDifference.', function() {
      var forward = agree('symdiff A,B', function(f) {
        return Array.from(f.symmetricDifference(new Set([1, 2, 3]), new Set([3, 4, 5])));
      });
      var backward = agree('symdiff B,A', function(f) {
        return Array.from(f.symmetricDifference(new Set([3, 4, 5]), new Set([1, 2, 3])));
      });

      assert.deepStrictEqual(forward, [1, 2, 4, 5]);
      assert.deepStrictEqual(backward, [4, 5, 1, 2]);
    });

    it('should add before deleting in disjunct.', function() {
      // {1,2} disjunct {2,3} is [1,3] and not [3,1] only because the addition
      // of B\A happens BEFORE the removal of the intersection.
      var order = agree('disjunct order', function(f) {
        var A = new Set([1, 2]);
        f.disjunct(A, new Set([2, 3]));
        return Array.from(A);
      });

      assert.deepStrictEqual(order, [1, 3]);
    });

    it('should not move a member that is re-added.', function() {
      // Set.add on a present member leaves it in place; only delete-then-add
      // moves it to the end. Both halves, because getting one right and the
      // other wrong is the likely mistake.
      var order = agree('re-add', function(f) {
        var A = new Set([1, 2, 3]);
        f.add(A, new Set([1]));
        return Array.from(A);
      });

      assert.deepStrictEqual(order, [1, 2, 3]);

      var moved = agree('delete then add', function(f) {
        var A = new Set([1, 2, 3]);
        f.subtract(A, new Set([1]));
        f.add(A, new Set([1]));
        return Array.from(A);
      });

      assert.deepStrictEqual(moved, [2, 3, 1]);
    });
  });

  describe('the mutating four', function() {

    ['add', 'subtract', 'intersect', 'disjunct'].forEach(function(name) {

      it('#.' + name + ' should mutate the caller\'s own object and return undefined.', function() {
        var A = new Set([1, 2]),
            before = A;

        var returned = port[name](A, new Set([2, 3]));

        assert.strictEqual(returned, undefined, name + ' returns nothing upstream');
        assert.strictEqual(A, before, name + ' must not replace the object');

        var theirs = new Set([1, 2]);
        upstream[name](theirs, new Set([2, 3]));

        assert.deepStrictEqual(Array.from(A), Array.from(theirs));
      });
    });

    it('should reach a live iterator through add/delete, not clear-and-rebuild.', function() {
      // THE spec of this file. A JS Set iterator is live: entries appended
      // after it was created are visited, and deleted ones are skipped. So a
      // bridge that emptied A and re-inserted every member would be observably
      // different here -- the iterator would see the re-inserted 1 as well --
      // while passing every assertion in test/set.js.
      var seen = agree('live iterator across add', function(f) {
        var A = new Set([1, 2]);
        var iterator = A.values();

        iterator.next();                       // consumes 1
        f.add(A, new Set([2, 3]));

        return Array.from(iterator);
      });

      assert.deepStrictEqual(seen, [2, 3]);
    });

    it('should reach a live iterator through disjunct\'s add-then-delete.', function() {
      var seen = agree('live iterator across disjunct', function(f) {
        var A = new Set([1, 2]);
        var iterator = A.values();

        iterator.next();                       // consumes 1
        f.disjunct(A, new Set([2, 3]));        // adds 3, then deletes 2

        return Array.from(iterator);
      });

      // 2 was deleted before the iterator reached it, so only 3 is left.
      assert.deepStrictEqual(seen, [3]);
    });

    it('should be defined when applied to their own argument.', function() {
      // Never done upstream, and every one of the four has a defined answer.
      // `subtract(A, A)` deletes from the set it is iterating and `disjunct`
      // both adds to and deletes from it.
      ['add', 'subtract', 'intersect', 'disjunct'].forEach(function(name) {
        var members = agree('self ' + name, function(f) {
          var A = new Set([1, 2, 3]);
          f[name](A, A);
          return Array.from(A);
        });

        assert.deepStrictEqual(
          members,
          name === 'add' || name === 'intersect' ? [1, 2, 3] : [],
          name + ' applied to itself');
      });
    });
  });

  describe('membership', function() {

    it('should treat NaN as one member, as SameValueZero does.', function() {
      var shared = agree('NaN membership', function(f) {
        return Array.from(f.intersection(new Set([NaN, 1]), new Set([NaN, 2])));
      });

      assert.strictEqual(shared.length, 1);
      assert.ok(Number.isNaN(shared[0]));
    });

    it('should treat -0 and 0 as one member, and store +0.', function() {
      var members = agree('negative zero', function(f) {
        return Array.from(f.union(new Set([-0]), new Set([0, 1])));
      });

      assert.deepStrictEqual(members, [0, 1]);
      assert.ok(!Object.is(members[0], -0), 'a Set stores -0 as +0');
    });

    it('should distinguish 1 from "1", as a Set does.', function() {
      var members = agree('no coercion', function(f) {
        return Array.from(f.union(new Set([1]), new Set(['1'])));
      });

      assert.deepStrictEqual(members, [1, '1']);
    });

    it('should reject object members with a message naming the limit.', function() {
      // A stated divergence (docs/modules/set.md): Set compares objects by
      // identity, and no identity hash for a JS object is reachable from Rust.
      // Upstream handles them; the port refuses, loudly. Asserted so that
      // silently starting to accept -- and therefore silently conflating two
      // distinct objects -- would be noticed.
      var A = new Set([{}]),
          B = new Set([{}]);

      assert.deepStrictEqual(Array.from(upstream.union(A, B)).length, 2);
      assert.throws(function() { port.union(A, B); }, /object/i);
    });
  });

  describe('arity', function() {

    it('should throw upstream\'s own message for a single argument.', function() {
      [['intersection', 'mnemonist/Set.intersection: needs at least two arguments.'],
       ['union', 'mnemonist/Set.union: needs at least two arguments.']].forEach(function(pair) {
        var name = pair[0], message = pair[1];

        [[new Set([1])], []].forEach(function(args) {
          var theirs = null, mine = null;

          try { upstream[name].apply(null, args); } catch (error) { theirs = error.message; }
          try { port[name].apply(null, args); } catch (error) { mine = error.message; }

          assert.strictEqual(theirs, message, 'upstream ' + name + '/' + args.length);
          assert.strictEqual(mine, message, 'port ' + name + '/' + args.length);
        });
      });
    });

    it('should accept more than four sets.', function() {
      var sets = [];
      for (var i = 0; i < 8; i++) sets.push(new Set([i, i + 1, 99]));

      var shared = agree('8-way intersection', function(f) {
        return Array.from(f.intersection.apply(null, sets));
      });
      var all = agree('8-way union', function(f) {
        return Array.from(f.union.apply(null, sets));
      });

      assert.deepStrictEqual(shared, [99]);
      assert.deepStrictEqual(all, [0, 1, 99, 2, 3, 4, 5, 6, 7, 8]);
    });
  });

  describe('the metrics', function() {

    it('should answer 0 rather than NaN for two empty sets.', function() {
      var answers = agree('empty metrics', function(f) {
        var empty = new Set();
        return [f.jaccard(empty, empty), f.overlap(empty, empty),
                f.intersectionSize(empty, empty), f.unionSize(empty, empty)];
      });

      assert.deepStrictEqual(answers, [0, 0, 0, 0]);
    });

    it('should agree with upstream on a set that is passed as both arguments.', function() {
      var answers = agree('self metrics', function(f) {
        var A = new Set([1, 2, 3]);
        return [f.jaccard(A, A), f.overlap(A, A),
                f.intersectionSize(A, A), f.unionSize(A, A),
                f.isSubset(A, A), f.isSuperset(A, A)];
      });

      assert.deepStrictEqual(answers, [1, 1, 3, 3, true, true]);
    });

    it('should agree with upstream on argument order for intersectionSize.', function() {
      // Upstream swaps so it walks the smaller set; the result must not depend
      // on which way round the caller passed them. test/set.js only ever
      // passes the larger first.
      var answers = agree('swapped intersectionSize', function(f) {
        var A = new Set([1, 2, 3, 4, 5]),
            B = new Set([4, 5, 6]);
        return [f.intersectionSize(A, B), f.intersectionSize(B, A),
                f.unionSize(A, B), f.unionSize(B, A),
                f.overlap(A, B), f.overlap(B, A)];
      });

      assert.deepStrictEqual(answers, [2, 2, 6, 6, 2 / 3, 2 / 3]);
    });
  });
});

File: crates/difffuzz/src/modules/set.rs
//! [`ModuleSpec`] for `set` — the second free-function module.
//!
//! The mode is the one `crate::modules::sort` introduced: no instance, no
//! observable state, and the comparison rests on the return value **plus every
//! argument after the call**. Here that second half is not a nicety — `add`,
//! `subtract`, `intersect` and `disjunct` all return `undefined` and do their
//! whole job to their first argument, so without the argument echo four of the
//! fourteen functions would be compared against nothing at all.
//!
//! # What the grammar reaches that `test/set.js` does not
//!
//! * **Order that is not the obvious order.** `intersection` iterates its
//!   *smallest* argument, so the result's order depends on which one that was,
//!   and ties go to the first. Sets here vary in size freely, so both branches
//!   occur constantly. The original file's two-set case uses equal sizes and
//!   its variadic case intersects down to a single member, where order is not
//!   visible.
//! * **Empty sets, everywhere.** `intersection` bails out on the first empty
//!   argument, `difference` short-circuits on either, and `jaccard`/`overlap`
//!   answer `0` rather than dividing. The original file has exactly one empty
//!   set and uses it in three of fourteen blocks.
//! * **Both arity throws.** `intersection` and `union` refuse fewer than two
//!   arguments with a specific message; the original file never calls either
//!   with one.
//! * **More than four sets**, and sets whose members are strings *and* numbers
//!   at once.
//! * **Repeat application.** A program is 1..80 calls, so a set that has been
//!   through `disjunct` is fed to `intersect`, and so on.
//!
//! # Deliberately excluded
//!
//! * **Object members.** `Set` compares them by identity and the bridge
//!   refuses them (`docs/modules/set.md`); the fuzzer works against core, whose
//!   member type is generic, so an object member is not even expressible here.
//! * **`NaN` and `-0` members.** They are SameValueZero's two special cases and
//!   they belong to the *bridge's* key type, not to core — `JsKey` normalises
//!   both, and core is generic over any `Hash + Eq`. Fuzzing them here would
//!   test `serde_json`'s round-tripping rather than the port.
//!   `tests/boundary/set.js` covers them where they live.
//! * **The three `===` shortcuts.** Upstream skips work when two arguments are
//!   the same *object*. The oracle decodes every `{"$set": …}` into a fresh
//!   `Set`, so the identity never holds on either side and the shortcuts are
//!   equally unreachable — no false divergence, and no coverage either.
//!   `tests/boundary/set.js` passes one object twice; core's own tests pass one
//!   reference twice.

use mnemonist_core::structures::set::{self as core_set, OrderedSet};
use proptest::collection;
use proptest::prelude::*;
use serde_json::{json, Value};

use crate::spec::{ModuleSpec, Op};

pub struct SetSpec;

/// Path proptest writes a minimised failing seed to.
pub const REGRESSIONS: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/proptest-regressions/set.txt");

/// The single upstream file this unit spans.
const FILES: &[&str] = &["set"];

/// The member pool, by index.
///
/// Small so that intersections are frequent rather than almost always empty,
/// and mixed so that `1` and `"1"` are two members — which they are to a `Set`,
/// and which a port that stringified its keys would get wrong.
const MEMBER_POOL: usize = 10;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
enum Member {
    Number(i64),
    Text(&'static str),
}

impl Member {
    fn at(index: usize) -> Self {
        match index {
            0 => Self::Number(0),
            1 => Self::Number(1),
            2 => Self::Number(2),
            3 => Self::Number(3),
            4 => Self::Number(4),
            5 => Self::Number(5),
            6 => Self::Text("1"),
            7 => Self::Text("a"),
            8 => Self::Text("b"),
            _ => Self::Text(""),
        }
    }

    fn to_json(&self) -> Value {
        match self {
            Self::Number(value) => json!(value),
            Self::Text(text) => json!(text),
        }
    }

    fn from_json(value: &Value) -> Self {
        match value {
            Value::Number(number) => {
                Self::Number(number.as_i64().expect("members are small integers"))
            }
            Value::String(text) => match text.as_str() {
                "1" => Self::Text("1"),
                "a" => Self::Text("a"),
                "b" => Self::Text("b"),
                "" => Self::Text(""),
                other => panic!("`{other}` is not a member this grammar generates"),
            },
            other => panic!("`{other}` is not a member this grammar generates"),
        }
    }
}

impl ModuleSpec for SetSpec {
    type Instance = ();

    fn module(&self) -> &'static str {
        "set"
    }

    fn functions(&self) -> &'static [&'static str] {
        FILES
    }

    /// None — see [`ModuleSpec::functions`] and the module docs.
    fn observations(&self) -> &'static [&'static str] {
        &[]
    }

    fn ctor_strategy(&self) -> BoxedStrategy<Vec<Value>> {
        Just(Vec::new()).boxed()
    }

    fn op_strategy(&self, _ctor: &[Value]) -> BoxedStrategy<Op> {
        prop_oneof![
            // The two variadic ones, from ONE set (which throws) to five.
            3 => sets(1..6).prop_map(|args| Op::new("intersection", args)),
            3 => sets(1..6).prop_map(|args| Op::new("union", args)),
            2 => sets(2..3).prop_map(|args| Op::new("difference", args)),
            2 => sets(2..3).prop_map(|args| Op::new("symmetricDifference", args)),
            2 => sets(2..3).prop_map(|args| Op::new("isSubset", args)),
            2 => sets(2..3).prop_map(|args| Op::new("isSuperset", args)),
            3 => sets(2..3).prop_map(|args| Op::new("add", args)),
            3 => sets(2..3).prop_map(|args| Op::new("subtract", args)),
            3 => sets(2..3).prop_map(|args| Op::new("intersect", args)),
            3 => sets(2..3).prop_map(|args| Op::new("disjunct", args)),
            2 => sets(2..3).prop_map(|args| Op::new("intersectionSize", args)),
            2 => sets(2..3).prop_map(|args| Op::new("unionSize", args)),
            2 => sets(2..3).prop_map(|args| Op::new("jaccard", args)),
            2 => sets(2..3).prop_map(|args| Op::new("overlap", args)),
        ]
        .boxed()
    }

    /// Shorter than the default: each op carries its own subjects rather than
    /// accumulating state in one instance.
    fn program_len(&self) -> std::ops::Range<usize> {
        1..80
    }

    fn construct(&self, _args: &[Value]) -> Self::Instance {}

    fn apply(&self, _instance: &mut Self::Instance, op: &Op) -> Value {
        let mut sets: Vec<OrderedSet<Member>> = op.args.iter().map(read).collect();

        // The arguments are echoed back from the PARSED sets, never from
        // `op.args`. The oracle re-encodes the real `Set` objects it built, so
        // a generated `{"$set": [1, 1, 2]}` comes back as `{"$set": [1, 2]}`;
        // echoing the raw argument would report a divergence on every
        // generated duplicate and none of them would be about the port.
        let echoed =
            |sets: &[OrderedSet<Member>]| -> Vec<Value> { sets.iter().map(write).collect() };

        match op.name {
            // The two variadic queries. Both can throw, and the throw is
            // compared like any other result.
            "intersection" | "union" => {
                let borrowed: Vec<&OrderedSet<Member>> = sets.iter().collect();
                let outcome = if op.name == "intersection" {
                    core_set::intersection(&borrowed)
                } else {
                    core_set::union(&borrowed)
                };

                let result = match outcome {
                    Ok(set) => write(&set),
                    Err(message) => json!({"$throw": message}),
                };

                returned(result, echoed(&sets))
            }

            "difference" => {
                let result = write(&core_set::difference(&sets[0], &sets[1]));

                returned(result, echoed(&sets))
            }
            "symmetricDifference" => {
                let result = write(&core_set::symmetric_difference(&sets[0], &sets[1]));

                returned(result, echoed(&sets))
            }
            "isSubset" => {
                let result = json!(core_set::is_subset(&sets[0], &sets[1]));

                returned(result, echoed(&sets))
            }
            "isSuperset" => {
                let result = json!(core_set::is_superset(&sets[0], &sets[1]));

                returned(result, echoed(&sets))
            }

            // The four mutators. `$return` is `undefined`; everything they do
            // is in the echoed first argument.
            "add" | "subtract" | "intersect" | "disjunct" => {
                let right = sets.pop().expect("two set arguments");
                let mut left = sets.pop().expect("two set arguments");

                match op.name {
                    "add" => core_set::add(&mut left, &right),
                    "subtract" => core_set::subtract(&mut left, &right),
                    "intersect" => core_set::intersect(&mut left, &right),
                    _ => core_set::disjunct(&mut left, &right),
                };

                returned(
                    json!({"$undefined": true}),
                    vec![write(&left), write(&right)],
                )
            }

            "intersectionSize" => {
                let result = json!(core_set::intersection_size(&sets[0], &sets[1]));

                returned(result, echoed(&sets))
            }
            "unionSize" => {
                let result = json!(core_set::union_size(&sets[0], &sets[1]));

                returned(result, echoed(&sets))
            }
            "jaccard" => {
                let result = number_json(core_set::jaccard(&sets[0], &sets[1]));

                returned(result, echoed(&sets))
            }
            "overlap" => {
                let result = number_json(core_set::overlap(&sets[0], &sets[1]));

                returned(result, echoed(&sets))
            }

            other => panic!("op `{other}` is not in this module's alphabet"),
        }
    }

    /// Nothing. See [`ModuleSpec::functions`].
    fn observe(&self, _instance: &mut Self::Instance) -> Value {
        json!({})
    }
}

/// `count` sets of 0..=8 members each, as `{"$set": […]}` arguments.
///
/// Sizes vary freely and independently, which is what makes `intersection`'s
/// smallest-argument rule -- and its tie-break -- reachable.
fn sets(count: std::ops::Range<usize>) -> BoxedStrategy<Vec<Value>> {
    collection::vec(collection::vec(0usize..MEMBER_POOL, 0..9), count)
        .prop_map(|sets| {
            sets.into_iter()
                .map(|members| {
                    let encoded: Vec<Value> = members
                        .into_iter()
                        .map(|index| Member::at(index).to_json())
                        .collect();

                    // Duplicates in the generated vector are legitimate: `new
                    // Set([1, 1])` is a one-member set, and dropping them here
                    // would stop the port's own de-duplication from being
                    // compared.
                    json!({"$set": encoded})
                })
                .collect()
        })
        .boxed()
}

/// Rebuild the [`OrderedSet`] a `{"$set": […]}` argument stands for.
fn read(argument: &Value) -> OrderedSet<Member> {
    let members = argument["$set"]
        .as_array()
        .expect("a set argument carries its members");

    OrderedSet::from_members(members.iter().map(Member::from_json))
}

/// A set, in the shape the oracle's `encode` renders one.
fn write(set: &OrderedSet<Member>) -> Value {
    let members: Vec<Value> = set.iter().map(Member::to_json).collect();

    json!({"$set": members})
}

/// The envelope `fuzz/oracle.js` wraps every free-function result in.
fn returned(result: Value, args: Vec<Value>) -> Value {
    json!({"$return": result, "$args": args})
}

/// A JavaScript number, encoded as `JSON.stringify` would encode it.
///
/// `jaccard` and `overlap` are the only functions here that can return a
/// non-integer, and both return the *integer* `0` on their zero shortcut.
/// serde_json compares `0` and `0.0` unequal, so emitting `json!(0.0)` there
/// would be a false divergence on every empty intersection. Duplicated from
/// `crate::modules::default_map`, which found the same trap on its first
/// campaign; shared code would be tidier and would put a merge conflict in a
/// file three worktrees edit.
fn number_json(value: f64) -> Value {
    if value.fract() == 0.0 && value.abs() < 9_007_199_254_740_992.0 {
        return json!(value as i64);
    }

    json!(value)
}
